File: dart/lib/src/rerout.dart
//
//  rerout
//  rerout.dart
//
//  Official Dart SDK for the Rerout branded-link API.
//
//  Created by Ngonidzashe Mangudya on 20/05/2026.
//  Copyright (c) 2026 Codecraft Solutions
//  Licensed under the MIT License
//  https://codecraftsolutions.co.za
//

import 'package:dio/dio.dart';
import 'package:rerout/src/models/batch_create_links.dart';
import 'package:rerout/src/models/create_link_request.dart';
import 'package:rerout/src/models/create_webhook_request.dart';
import 'package:rerout/src/models/created_webhook.dart';
import 'package:rerout/src/models/link_stats.dart';
import 'package:rerout/src/models/list_links_result.dart';
import 'package:rerout/src/models/list_webhooks_result.dart';
import 'package:rerout/src/models/project_stats.dart';
import 'package:rerout/src/models/qr_options.dart';
import 'package:rerout/src/models/record_conversion_request.dart';
import 'package:rerout/src/models/recorded_conversion.dart';
import 'package:rerout/src/models/rerout_exception.dart';
import 'package:rerout/src/models/result.dart';
import 'package:rerout/src/models/short_link.dart';
import 'package:rerout/src/models/update_link_request.dart';

/// Default production API URL.
const String kReroutDefaultBaseUrl = 'https://api.rerout.co';

/// Official Dart SDK for the Rerout branded-link API.
///
/// ## Usage
///
/// ```dart
/// import 'package:rerout/rerout.dart';
///
/// final rerout = Rerout.initialize(apiKey: 'rrk_...');
///
/// final result = await rerout.links.create(
///   const CreateLinkRequest(
///     targetUrl: 'https://example.com/q4-sale',
///     domainHostname: 'go.brand.com',
///     code: 'q4',
///   ),
/// );
///
/// switch (result) {
///   case Success(:final data):
///     print('Short URL: ${data.shortUrl}');
///   case Error(:final error):
///     print('Failed: ${error.message}');
/// }
/// ```
class Rerout {
  Rerout._({
    required String apiKey,
    required String baseUrl,
    required Dio dio,
  }) : _apiKey = apiKey,
       _baseUrl = baseUrl,
       _dio = dio {
    links = Links._(this);
    project = Project._(this);
    qr = Qr._(this);
    webhooks = Webhooks._(this);
    conversions = Conversions._(this);
  }

  /// Creates a new instance of the Rerout SDK.
  ///
  /// ## Parameters
  ///
  /// - [apiKey]: Project API key from the dashboard (`rrk_…`). Required.
  /// - [baseUrl]: Override the API base URL. Defaults to
  ///   `https://api.rerout.co`. Useful for staging or self-hosted setups.
  /// - [dio]: Inject a pre-configured Dio instance. Useful for tests, custom
  ///   interceptors, proxies, etc.
  ///
  /// ## Example
  ///
  /// ```dart
  /// final rerout = Rerout.initialize(apiKey: 'rrk_...');
  /// ```
  factory Rerout.initialize({
    required String apiKey,
    String baseUrl = kReroutDefaultBaseUrl,
    Dio? dio,
  }) {
    if (apiKey.trim().isEmpty) {
      throw const ReroutException(
        code: 'missing_api_key',
        message: 'A project API key is required to construct Rerout.',
        statusCode: 0,
      );
    }
    return Rerout._(
      apiKey: apiKey,
      baseUrl: baseUrl.replaceAll(RegExp(r'/+$'), ''),
      dio: dio ?? Dio(),
    );
  }

  final String _apiKey;
  final String _baseUrl;
  final Dio _dio;

  /// Link operations: create, list, get, update, delete, stats.
  late final Links links;

  /// Project-level operations: aggregate stats, current project.
  late final Project project;

  /// QR helpers — URL builders and signed-fetch.
  late final Qr qr;

  /// Webhook endpoint management: create, list, delete.
  late final Webhooks webhooks;

  /// Conversion tracking: record a conversion against a prior click.
  late final Conversions conversions;

  /// The resolved API base URL. Exposed for diagnostics and the QR helper.
  String get baseUrl => _baseUrl;

  /// @internal — used by [Links], [Project], [Qr].
  Future<Result<T>> request<T>({
    required String method,
    required String path,
    required T Function(dynamic data) parse,
    Map<String, dynamic>? body,
    Map<String, dynamic>? query,
    ResponseType responseType = ResponseType.json,
  }) async {
    final url = '$_baseUrl$path';
    try {
      final response = await _dio.request<dynamic>(
        url,
        data: body,
        queryParameters: query,
        options: Options(
          method: method,
          responseType: responseType,
          headers: <String, String>{
            'authorization': 'Bearer $_apiKey',
            'accept': responseType == ResponseType.plain
                ? 'image/svg+xml,text/html'
                : 'application/json',
            if (body != null) 'content-type': 'application/json',
          },
          validateStatus: (status) => status != null && status < 600,
        ),
      );
      final status = response.statusCode ?? 0;
      if (status >= 200 && status < 300) {
        return Result.success(parse(response.data));
      }
      return Result.error(_errorFromResponse(response, path));
    } on DioException catch (e) {
      return Result.error(_errorFromDio(e, path));
    } on Exception catch (e) {
      return Result.error(
        ReroutException(
          code: 'client_error',
          message: e.toString(),
          statusCode: 0,
          path: path,
        ),
      );
    }
  }

  ReroutException _errorFromResponse(Response<dynamic> response, String path) {
    final status = response.statusCode ?? 0;
    final data = response.data;
    String? code;
    String? message;
    String? timestamp;
    if (data is Map<String, dynamic>) {
      code = data['code'] as String?;
      message = data['message'] as String?;
      timestamp = data['timestamp'] as String?;
    }
    return ReroutException(
      code: code ?? _synthCodeForStatus(status),
      message: message ?? 'Rerout returned HTTP $status.',
      statusCode: status,
      path: path,
      timestamp: timestamp,
    );
  }

  ReroutException _errorFromDio(DioException e, String path) {
    final response = e.response;
    if (response != null) {
      return _errorFromResponse(response, path);
    }
    final isTimeout =
        e.type == DioExceptionType.connectionTimeout ||
        e.type == DioExceptionType.receiveTimeout ||
        e.type == DioExceptionType.sendTimeout;
    return ReroutException(
      code: isTimeout ? 'timeout' : 'network_error',
      message: e.message ?? 'Request to Rerout failed.',
      statusCode: 0,
      path: path,
    );
  }

  static String _synthCodeForStatus(int status) {
    if (status == 401) return 'unauthorized';
    if (status == 403) return 'forbidden';
    if (status == 404) return 'not_found';
    if (status == 429) return 'rate_limited';
    if (status >= 500 && status < 600) return 'server_error';
    return 'client_error';
  }
}

// ─── Namespaces ─────────────────────────────────────────────────────────────

/// Link operations namespace. Reached via [Rerout.links].
class Links {
  Links._(this._client);

  final Rerout _client;

  /// Creates a new short link.
  Future<Result<ShortLink>> create(CreateLinkRequest request) {
    return _client.request<ShortLink>(
      method: 'POST',
      path: '/v1/links',
      body: request.toJson(),
      parse: (data) => ShortLink.fromJson(data as Map<String, dynamic>),
    );
  }

  /// Bulk-creates links in one request. Each item succeeds or fails
  /// independently; inspect `results[i].ok` for per-item outcomes.
  Future<Result<BatchCreateLinksResult>> createBatch(
    List<BatchLinkInput> links,
  ) {
    return _client.request<BatchCreateLinksResult>(
      method: 'POST',
      path: '/v1/links/batch',
      body: {'links': links.map((l) => l.toJson()).toList()},
      parse: (data) =>
          BatchCreateLinksResult.fromJson(data as Map<String, dynamic>),
    );
  }

  /// Paginated list of links in the project.
  Future<Result<ListLinksResult>> list({int? cursor, int? limit}) {
    final query = <String, dynamic>{};
    if (cursor != null) query['cursor'] = cursor;
    if (limit != null) query['limit'] = limit;
    return _client.request<ListLinksResult>(
      method: 'GET',
      path: '/v1/links',
      query: query.isEmpty ? null : query,
      parse: (data) => ListLinksResult.fromJson(data as Map<String, dynamic>),
    );
  }

  /// Get a single link by code.
  Future<Result<ShortLink>> get(String code) {
    return _client.request<ShortLink>(
      method: 'GET',
      path: '/v1/links/${Uri.encodeComponent(code)}',
      parse: (data) => ShortLink.fromJson(data as Map<String, dynamic>),
    );
  }

  /// Patch a link. Only fields set on [request] are sent.
  Future<Result<ShortLink>> update(String code, UpdateLinkRequest request) {
    if (request.isEmpty) {
      return Future.value(
        Result.error(
          ReroutException(
            code: 'bad_request',
            message: 'UpdateLinkRequest has no fields to send.',
            statusCode: 0,
            path: '/v1/links/$code',
          ),
        ),
      );
    }
    return _client.request<ShortLink>(
      method: 'PATCH',
      path: '/v1/links/${Uri.encodeComponent(code)}',
      body: request.toJson(),
      parse: (data) => ShortLink.fromJson(data as Map<String, dynamic>),
    );
  }

  /// Soft-deletes a link. The short URL stops redirecting and is gone from
  /// lists.
  Future<Result<bool>> delete(String code) {
    return _client.request<bool>(
      method: 'DELETE',
      path: '/v1/links/${Uri.encodeComponent(code)}',
      parse: (data) {
        if (data is Map<String, dynamic>) {
          return data['deleted'] as bool? ?? true;
        }
        return true;
      },
    );
  }

  /// Per-link click stats. Defaults to 30 days.
  Future<Result<LinkStats>> stats(String code, {int days = 30}) {
    return _client.request<LinkStats>(
      method: 'GET',
      path: '/v1/links/${Uri.encodeComponent(code)}/stats',
      query: {'days': days},
      parse: (data) => LinkStats.fromJson(data as Map<String, dynamic>),
    );
  }
}

/// Project-level operations namespace. Reached via [Rerout.project].
class Project {
  Project._(this._client);

  final Rerout _client;

  /// Aggregate stats across every link in the project.
  Future<Result<ProjectStats>> stats({int days = 30}) {
    return _client.request<ProjectStats>(
      method: 'GET',
      path: '/v1/projects/me/stats',
      query: {'days': days},
      parse: (data) => ProjectStats.fromJson(data as Map<String, dynamic>),
    );
  }

  /// Returns info about the project that owns the current API key.
  Future<Result<Map<String, dynamic>>> me() {
    return _client.request<Map<String, dynamic>>(
      method: 'GET',
      path: '/v1/projects/me',
      parse: (data) =>
          (data as Map<String, dynamic>?) ?? const <String, dynamic>{},
    );
  }
}

/// Webhook endpoint management namespace. Reached via [Rerout.webhooks].
///
/// Manages the project's *outbound* webhook endpoints. To verify *inbound*
/// deliveries, use `ReroutWebhookSignature` instead.
class Webhooks {
  Webhooks._(this._client);

  final Rerout _client;

  /// Creates a webhook endpoint for the project that owns the API key.
  ///
  /// The returned [CreatedWebhook.signingSecret] is shown once — persist it to
  /// verify inbound deliveries.
  Future<Result<CreatedWebhook>> create(CreateWebhookRequest request) {
    return _client.request<CreatedWebhook>(
      method: 'POST',
      path: '/v1/projects/me/webhooks',
      body: request.toJson(),
      parse: (data) =>
          CreatedWebhook.fromJson(data as Map<String, dynamic>),
    );
  }

  /// Lists webhook endpoints and the event types the server can deliver.
  Future<Result<ListWebhooksResult>> list() {
    return _client.request<ListWebhooksResult>(
      method: 'GET',
      path: '/v1/projects/me/webhooks',
      parse: (data) =>
          ListWebhooksResult.fromJson(data as Map<String, dynamic>),
    );
  }

  /// Soft-deletes an endpoint and abandons its pending deliveries.
  /// Idempotent.
  Future<Result<bool>> delete(String endpointId) {
    return _client.request<bool>(
      method: 'DELETE',
      path: '/v1/projects/me/webhooks/${Uri.encodeComponent(endpointId)}',
      parse: (data) {
        if (data is Map<String, dynamic>) {
          return data['deleted'] as bool? ?? true;
        }
        return true;
      },
    );
  }
}

/// Conversion tracking namespace. Reached via [Rerout.conversions].
class Conversions {
  Conversions._(this._client);

  final Rerout _client;

  /// Records a conversion attributed to a prior click via its `clickId`
  /// (`rrid`). Idempotent per `(clickId, eventName)`.
  Future<Result<RecordedConversion>> record(RecordConversionRequest request) {
    return _client.request<RecordedConversion>(
      method: 'POST',
      path: '/v1/conversions',
      body: request.toJson(),
      parse: (data) =>
          RecordedConversion.fromJson(data as Map<String, dynamic>),
    );
  }
}

/// QR helpers — URL builders and signed-fetch. Reached via [Rerout.qr].
class Qr {
  Qr._(this._client);

  final Rerout _client;

  /// Builds the URL the Rerout API serves the QR SVG from. Pure — does not
  /// call the API.
  ///
  /// ```dart
  /// final url = rerout.qr.url(
  ///   'q4',
  ///   options: const QrOptions(size: 12, ecc: 'H'),
  /// );
  /// ```
  String url(String code, {QrOptions? options}) {
    final base = _client.baseUrl.replaceAll(RegExp(r'/+$'), '');
    final uri = Uri.parse('$base/v1/links/${Uri.encodeComponent(code)}/qr');
    final params = options?.toQueryParameters() ?? const <String, String>{};
    if (params.isEmpty) return uri.toString();
    return uri.replace(queryParameters: params).toString();
  }

  /// Fetches the QR SVG body. Hits the same endpoint as [url] but attaches
  /// the bearer token and returns the rendered text.
  Future<Result<String>> svg(String code, {QrOptions? options}) {
    return _client.request<String>(
      method: 'GET',
      path: '/v1/links/${Uri.encodeComponent(code)}/qr',
      query: options?.toQueryParameters(),
      responseType: ResponseType.plain,
      parse: (data) => data is String ? data : data.toString(),
    );
  }
}

File: php/src/Rerout.php
<?php

declare(strict_types=1);

namespace Rerout;

use GuzzleHttp\Client;
use GuzzleHttp\ClientInterface;
use GuzzleHttp\Exception\ConnectException;
use GuzzleHttp\Exception\GuzzleException;
use GuzzleHttp\Exception\RequestException;
use GuzzleHttp\Exception\TransferException;
use GuzzleHttp\Psr7\Request as Psr7Request;
use JsonException;
use Psr\Http\Message\ResponseInterface;
use Rerout\Exceptions\ReroutException;
use Rerout\Resources\Conversions;
use Rerout\Resources\Links;
use Rerout\Resources\Project;
use Rerout\Resources\Qr;
use Rerout\Resources\Tags;
use Rerout\Resources\Webhooks;

/**
 * Official PHP client for the Rerout branded-link API.
 *
 * ## Usage
 *
 * ```php
 * use Rerout\Rerout;
 * use Rerout\Models\CreateLinkInput;
 *
 * $rerout = new Rerout(getenv('REROUT_API_KEY'));
 *
 * $link = $rerout->links()->create(new CreateLinkInput(
 *     targetUrl: 'https://example.com/q4-sale',
 *     domainHostname: 'go.brand.com',
 *     code: 'q4',
 * ));
 *
 * echo $link->shortUrl;
 * ```
 *
 * Construction options (passed as `$options` array):
 *
 * - `base_url` — override the API base URL. Defaults to `https://api.rerout.co`.
 * - `timeout` — request timeout in seconds. Defaults to 30.
 * - `client` — inject a pre-configured `\GuzzleHttp\ClientInterface` (e.g.
 *   for tests with a `MockHandler`).
 * - `default_headers` — extra headers appended to every request.
 */
final class Rerout
{
    /** Default production API URL. Override via the `base_url` option. */
    public const string DEFAULT_BASE_URL = 'https://api.rerout.co';

    /** Default per-request timeout in seconds. */
    public const int DEFAULT_TIMEOUT_SECONDS = 30;

    private readonly string $apiKey;
    private readonly string $baseUrl;
    private readonly int $timeout;
    private readonly ClientInterface $http;
    /** @var array<string, string> */
    private readonly array $defaultHeaders;

    private readonly Links $links;
    private readonly Project $project;
    private readonly Qr $qr;
    private readonly Webhooks $webhooks;
    private readonly Tags $tags;
    private readonly Conversions $conversions;

    /**
     * @param string                                                                              $apiKey  Project API key (`rrk_…`). Required.
     * @param array{base_url?: string, timeout?: int, client?: ClientInterface, default_headers?: array<string, string>} $options Construction options.
     */
    public function __construct(string $apiKey, array $options = [])
    {
        if (trim($apiKey) === '') {
            throw new ReroutException(
                errorCode: 'missing_api_key',
                message: 'A project API key is required to construct Rerout.',
                status: 0,
            );
        }
        $this->apiKey = $apiKey;

        $baseUrl = $options['base_url'] ?? self::DEFAULT_BASE_URL;
        $this->baseUrl = rtrim($baseUrl, '/');

        $this->timeout = $options['timeout'] ?? self::DEFAULT_TIMEOUT_SECONDS;

        $this->http = $options['client'] ?? new Client();

        $this->defaultHeaders = $options['default_headers'] ?? [];

        $this->links = new Links($this);
        $this->project = new Project($this);
        $this->qr = new Qr($this);
        $this->webhooks = new Webhooks($this);
        $this->tags = new Tags($this);
        $this->conversions = new Conversions($this);
    }

    /** Link operations: create, list, get, update, delete, stats. */
    public function links(): Links
    {
        return $this->links;
    }

    /** Project-level operations: aggregate stats, current project. */
    public function project(): Project
    {
        return $this->project;
    }

    /** QR helpers — URL builders and signed-fetch. */
    public function qr(): Qr
    {
        return $this->qr;
    }

    /** Webhook endpoint management: create, list, delete. */
    public function webhooks(): Webhooks
    {
        return $this->webhooks;
    }

    /** Tag management: list, create, update, delete. */
    public function tags(): Tags
    {
        return $this->tags;
    }

    /** Conversion tracking: record conversion events against a click. */
    public function conversions(): Conversions
    {
        return $this->conversions;
    }

    /** The resolved API base URL (trailing slashes stripped). */
    public function baseUrl(): string
    {
        return $this->baseUrl;
    }

    /**
     * @internal Used by {@see Links}, {@see Project}, {@see Qr}.
     *
     * Sends a request to the Rerout API and returns the decoded JSON body.
     * On 2xx with an empty body returns `null`. On error throws a
     * {@see ReroutException}. Set `$expectJson` to `false` to return the
     * raw response body as a string (used by the QR SVG fetch).
     *
     * @param array<string, scalar|null> $query
     * @param array<string, mixed>|null $body
     *
     * @return array<string, mixed>|string|null
     */
    public function request(
        string $method,
        string $path,
        array $query = [],
        ?array $body = null,
        bool $expectJson = true,
    ): array|string|null {
        $url = $this->baseUrl . $path;
        $filteredQuery = [];
        foreach ($query as $key => $value) {
            if ($value !== null) {
                $filteredQuery[$key] = (string) $value;
            }
        }
        if ($filteredQuery !== []) {
            $url .= '?' . http_build_query($filteredQuery, '', '&', PHP_QUERY_RFC3986);
        }

        $headers = [
            'Authorization' => 'Bearer ' . $this->apiKey,
            'Accept' => $expectJson ? 'application/json' : 'image/svg+xml,text/html',
        ];
        foreach ($this->defaultHeaders as $name => $value) {
            $headers[$name] = $value;
        }

        $bodyString = null;
        if ($body !== null) {
            try {
                $bodyString = json_encode($body, JSON_THROW_ON_ERROR | JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE);
            } catch (JsonException $e) {
                throw new ReroutException(
                    errorCode: 'client_error',
                    message: 'Failed to encode request body as JSON: ' . $e->getMessage(),
                    status: 0,
                    path: $path,
                    previous: $e,
                );
            }
            $headers['Content-Type'] = 'application/json';
        }

        $request = new Psr7Request($method, $url, $headers, $bodyString);

        try {
            $response = $this->http->send($request, [
                'timeout' => $this->timeout,
                'connect_timeout' => $this->timeout,
                'http_errors' => false,
            ]);
        } catch (ConnectException $e) {
            throw new ReroutException(
                errorCode: self::isTimeout($e) ? 'timeout' : 'network_error',
                message: $e->getMessage(),
                status: 0,
                path: $path,
                previous: $e,
            );
        } catch (RequestException $e) {
            $resp = $e->getResponse();
            if ($resp !== null) {
                throw $this->errorFromResponse($resp, $path, $e);
            }

            throw new ReroutException(
                errorCode: self::isTimeout($e) ? 'timeout' : 'network_error',
                message: $e->getMessage(),
                status: 0,
                path: $path,
                previous: $e,
            );
        } catch (TransferException $e) {
            throw new ReroutException(
                errorCode: self::isTimeout($e) ? 'timeout' : 'network_error',
                message: $e->getMessage(),
                status: 0,
                path: $path,
                previous: $e,
            );
        } catch (GuzzleException $e) {
            throw new ReroutException(
                errorCode: 'network_error',
                message: $e->getMessage(),
                status: 0,
                path: $path,
                previous: $e,
            );
        }

        return $this->handleResponse($response, $path, $expectJson);
    }

    /**
     * @return array<string, mixed>|string|null
     */
    private function handleResponse(ResponseInterface $response, string $path, bool $expectJson): array|string|null
    {
        $status = $response->getStatusCode();
        $body = (string) $response->getBody();

        if ($status < 200 || $status >= 300) {
            throw $this->errorFromBody($status, $body, $path);
        }

        if (!$expectJson) {
            return $body;
        }

        if ($body === '') {
            return null;
        }

        try {
            /** @var array<string, mixed>|mixed $decoded */
            $decoded = json_decode($body, true, 512, JSON_THROW_ON_ERROR);
        } catch (JsonException $e) {
            throw new ReroutException(
                errorCode: 'unexpected_response',
                message: 'Rerout returned a non-JSON success body.',
                status: $status,
                path: $path,
                details: ['body' => $body],
                previous: $e,
            );
        }

        if (!is_array($decoded)) {
            throw new ReroutException(
                errorCode: 'unexpected_response',
                message: 'Rerout returned a non-object JSON success body.',
                status: $status,
                path: $path,
                details: ['body' => $body],
            );
        }

        /** @var array<string, mixed> $decoded */

        return $decoded;
    }

    private function errorFromResponse(ResponseInterface $response, string $path, ?\Throwable $previous = null): ReroutException
    {
        return $this->errorFromBody(
            $response->getStatusCode(),
            (string) $response->getBody(),
            $path,
            $previous,
        );
    }

    private function errorFromBody(int $status, string $body, string $path, ?\Throwable $previous = null): ReroutException
    {
        if ($body === '') {
            return new ReroutException(
                errorCode: self::synthCodeForStatus($status),
                message: sprintf('Rerout returned HTTP %d with no body.', $status),
                status: $status,
                path: $path,
                previous: $previous,
            );
        }

        $code = null;
        $message = null;
        $timestamp = null;
        $details = null;

        try {
            /** @var mixed $decoded */
            $decoded = json_decode($body, true, 512, JSON_THROW_ON_ERROR);
            if (is_array($decoded)) {
                /** @var array<string, mixed> $decoded */
                $details = $decoded;
                $rawCode = $decoded['code'] ?? null;
                if (is_string($rawCode)) {
                    $code = $rawCode;
                }
                $rawMessage = $decoded['message'] ?? null;
                if (is_string($rawMessage)) {
                    $message = $rawMessage;
                }
                $rawTimestamp = $decoded['timestamp'] ?? null;
                if (is_string($rawTimestamp)) {
                    $timestamp = $rawTimestamp;
                }
            }
        } catch (JsonException) {
            // Fall through and treat as a non-JSON error body.
        }

        return new ReroutException(
            errorCode: $code ?? self::synthCodeForStatus($status),
            message: $message ?? sprintf('Rerout returned HTTP %d.', $status),
            status: $status,
            path: $path,
            timestamp: $timestamp,
            details: $details,
            previous: $previous,
        );
    }

    private static function isTimeout(\Throwable $e): bool
    {
        $msg = strtolower($e->getMessage());

        return str_contains($msg, 'timed out') || str_contains($msg, 'timeout');
    }

    private static function synthCodeForStatus(int $status): string
    {
        return match (true) {
            $status === 401 => 'unauthorized',
            $status === 403 => 'forbidden',
            $status === 404 => 'not_found',
            $status === 429 => 'rate_limited',
            $status >= 500 && $status < 600 => 'server_error',
            default => 'client_error',
        };
    }
}
